docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,643 @@
+
+### Block Matrix Multiplication
+
+Let's implement `matmul(x, y)` function that multiplies an `(m, k)` array with a `(k, n)` array, but let's say that we also have a function `matmul_small` that does matrix multiplication for us but only with small sizes (say `m, k, n <= 256`). Could we implement our our `matmul` function in terms of `matmul_small`?
+
+Yes! The answer is by decomposing the matmul into one that operates on small chunks of our inputs (a.k.a. blocks).
+
+<!-- Formally, if we have input arrays $x \in \mathbb{R}^{m \times k}$ and $y \in \mathbb{R}^{k \times n}$ and output $z \in \mathbb{R}^{m \times n}$, we decompose them into blocks along the dimensions of size $b_m, b_k, b_n$.
+
+$$
+\begin{bmatrix}
+x_{0, 0} & \cdots & x_{0, i_k} \\
+x_{1, 0} & \cdots & x_{1, i_k} \\
+\vdots & \ddots & \vdots \\
+x_{i_m, 0} & \cdots & x_{i_m, i_k} \\
+\end{bmatrix}
+$$
+ where $i_m \in [0, \ldots, B_m]$ -->
+
+Here's a NumPy implementation:
+
+```{code-cell} ipython3
+:id: PACqDMtQrMOL
+
+def matmul_small(x: np.ndarray, y: np.ndarray) -> np.ndarray:
+  m, k, n = x.shape[0], x.shape[1], y.shape[0]
+  assert m <= 256
+  assert k <= 256
+  assert n <= 256
+  return np.matmul(x, y)
+
+def block_matmul(
+    x: np.ndarray,
+    y: np.ndarray,
+    *,
+    bm: int = 256,
+    bk: int = 256,
+    bn: int = 256,
+) -> np.ndarray:
+  m, k = x.shape
+  _, n = y.shape
+
+  z = np.zeros((m, n), dtype=x.dtype)
+  for m_i in range(m // bm):
+    for n_i in range(n // bn):
+      for k_i in range(k // bk):
+        m_slice = slice(m_i * bm, (m_i + 1) * bm)
+        k_slice = slice(k_i * bk, (k_i + 1) * bk)
+        n_slice = slice(n_i * bn, (n_i + 1) * bn)
+        x_block = x[m_slice, k_slice]
+        y_block = y[k_slice, n_slice]
+        z[m_slice, n_slice] += matmul_small(x_block, y_block)
+  return z
+```
+
++++ {"id": "TP49TV6q8so9"}
+
+Our `block_matmul` function should now work on inputs larger than 256.
+
+```{code-cell} ipython3
+:id: 2SZFnWnurzEC
+
+m, k, n = 4096, 4096, 4096
+x = np.random.uniform(size=(m, k)).astype(np.float32)
+y = np.random.uniform(size=(k, n)).astype(np.float32)
+np.testing.assert_allclose(x @ y, block_matmul(x, y), atol=1e-6, rtol=1e-6)
+```
+
++++ {"id": "GXtjEtEhtARN"}
+
+`block_matmul` decomposes a matrix multiplication into many smaller ones by observing that each output chunk of size `(bm, bn)` can be computed by accumulating several `(bm, bk) x (bk, bn)` size matrix multiplications.
+
+TPUs have native hardware capable of small matrix multiplication akin to `matmul_small`, so to utilize this hardware when doing bigger matrix multiplications, we will apply the `block_matmul` decomposition.
+
++++ {"id": "a0ESFoX1ID0z"}
+
+### Pipelining
+
+In [the previous guide](pipelining), we covered how pipelining in Pallas works. To make sure our compute units are always working and never waiting for memory transfers, we overlap the memory transfers for the next iteration of a kernel with the current one.
```

Comment:
waiting -> stalled by the memory transfers? Technically the MXU never waits on memory